src/inference.rs

```diff
@@ -73,6 +73,24 @@ fn unify(mut x: RcVar, mut y: RcVar, hint: &'static str) -> Result<(), Error> {
         VariableInner::Free(_) => Ok(()),
         // If y was already bound to a type, then x must be bound, too
         VariableInner::Bound(y_ty, _) => bind(&x, y_ty, hint),
+        // If y was precomputed, same story, but in this case we cannot modify `y` (since it
+        // needs to be shared across many programs, and setting it to `EqualTo` something
+        // local to a single program would cause inference failures), so put its old value back.
+        //
+        // This means that rather than x being the parent of y, y remains as its own parent,
+        // disconnected from x. This leaves two equivalence classes where there should be one:
+        // the one in which y is the root, and the one which includes x. By calling `bind` on
+        // x, we ensure that both classes actually have the same concrete type.
+        //
+        // Future changes to one class will not propagate to the other, but this is OK *as long
+        // as* all precomputed types are concrete types; then there are no changes to propagate
+        // and any other variables that get unified into either class will simply have their
+        // own type concretized into this one.
+        VariableInner::Precomputed(y_ty, y_var) => {
+            bind(&x, y_ty.clone(), hint)?;
+            y.borrow_mut().inner = VariableInner::Precomputed(y_ty, y_var);
+            Ok(())
+        }
```

Comment:
Also the identifier `y_var` is confusing because it is of type `Type`. Maybe `y_final_ty`. The other `*_ty` variables are implicitly variable types.

Reply:
> If y was already bound to a precomputed type, then x must be bound, too. During binding, variables are replaced with their root.

They are not replaced with their root, but replaced with a (possibly indirect) reference to their root (i.e. put into an equivalence class with the root). Undoing this means that you are left with two "equivalent" variables that are not actually connected, which justifies the long comment.

> Also the identifier `y_var` is confusing because it is of type `Type`. Maybe `y_final_ty`. The other `*_ty` variables are implicitly variable types.

Lol, this is true. I'll fix it.

Reply:
Ok, `y` is replaced with something else while `y_ty` and `y_var` remain constant. We want `y` to remain constant, too, so we put its old value back. IDK if letting it vary would cause any issues, but it could lead to weird race conditions. If `x` cannot be unified with `y`, leading to an error, that wouldn't be good for following runs of the inference algorithm.

The comment makes more sense now.

Reply:
It will cause issues. Remove the code and try to run the tests :)